pallets/token-factory/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+			let address = <ContractAddress<T>>::get(&id).ok_or(Error::<T>::IdNotClaimed)?;
+			// check that `who` has at least `amount` of `id` token
+			let balance = Self::balance_of(id, who)?;
+			let amount_to_burn = if amount > balance { balance } else { amount };
+			ensure!(
+				amount_to_burn > T::Balance::zero(),
+				Error::<T>::NotEnoughBalanceToBurn
+			);
+			let mut nonce = <Nonce<T>>::get();
+			let mut input = hex_literal::hex!("9dc29fac").to_vec();
+			// append address
+			input.extend_from_slice(H256::from(who.clone()).as_bytes());
+			// append amount
+			input.extend_from_slice(
+				H256::from_uint(&U256::from(amount_to_burn.saturated_into::<u128>())).as_bytes(),
+			);
+			match T::Runner::call(
+				// source: H160
+				Self::sudo_caller(),
+				// target
+				address,
+				// input: Vec<u8>
+				input,
+				// value: U256
+				U256::zero(),
+				// gas limit: U256
+				U256::from(0x10000000).low_u64(),
+				// gas price: U256
+				Some(U256::from(1)),
+				// nonce: Option<H256>
+				Some(nonce),
+				// config: EvmConfig
+				T::config(),
+			)? {
+				ExecutionInfo {
+					exit_reason: ExitReason::Succeed(_),
+					..
+				} => {
+					Self::deposit_event(Event::Burned(
+						id,
+						T::AddressMapping::into_account_id(who),
+						amount_to_burn,
+					));
+				}
+				ExecutionInfo {
+					exit_reason: reason,
+					..
+				} => {
+					Self::deposit_event(Event::EvmCallFailed(EvmCall::Burn, reason));
+				}
+			}
+			// increment nonce
+			nonce += U256::from(1);
+			<Nonce<T>>::put(nonce);
+			Ok(().into())
+		}
+		/// Gets total issuance for the given token if it exists in local evm instance
+		fn total_issuance(id: T::TokenId) -> Result<T::Balance, DispatchError> {
+			let address = <ContractAddress<T>>::get(id).ok_or(Error::<T>::IdNotClaimed)?;
+			let mut nonce = <Nonce<T>>::get();
+			// first 4 bytes of hex output of Sha3("totalSupply()")
+			let input = hex_literal::hex!("18160ddd").to_vec();
+			match T::Runner::call(
+				// source: H160
+				Self::sudo_caller(),
+				// target
+				address,
+				// input: Vec<u8>
+				input,
+				// value: U256
+				U256::zero(),
+				// gas limit: U256
+				U256::from(0x10000000).low_u64(),
+				// gas price: U256
+				Some(U256::from(1)),
+				// nonce: Option<H256>
+				Some(nonce),
+				// config: EvmConfig
+				T::config(),
+			) {
+				Ok(ExecutionInfo {
+					exit_reason: ExitReason::Succeed(_),
+					value: result,
+					..
+				}) => {
+					// increment nonce
+					nonce += U256::from(1);
+					<Nonce<T>>::put(nonce);
+					let value = U256::from(result.as_slice()).saturated_into::<u128>();
+					Ok(value.saturated_into::<T::Balance>())
+				}
+				Ok(ExecutionInfo {
+					exit_reason: reason,
+					..
+				}) => {
+					// increment nonce
+					nonce += U256::from(1);
+					<Nonce<T>>::put(nonce);
+					Self::deposit_event(Event::EvmCallFailed(EvmCall::TotalIssuance, reason));
+					Ok(T::Balance::zero())
+				}
+				Err(e) => Err(e.into()),
```

Comment:
We don't need to increment the nonce if the EVM call fails. Right? Might need to test this assumption.

Reply:
If we ever forget to increment the nonce, we'll get the `InvalidNonce` error for all subsequent calls...

Reply:
We don't, the current tests make failed calls to the EVM without incrementing the nonce and subsequent calls succeed